dhis-2/dhis-e2e-test/src/test/java/org/hisp/dhis/tracker/acl/TrackedEntityInstanceAclReadTests.java

```diff
@@ -0,0 +1,539 @@
+                    if ( object.get( "publicAccess" ).getAsString().matches( "..r.*" ) )
+                    {
+                        hasDataRead = true;
+                    }
+                    else
+                    {
+                        JsonArray userAccesses = object.getAsJsonArray( "userAccesses" ).getAsJsonArray();
+                        JsonArray userGroupAccess = object.getAsJsonArray( "userGroupAccesses" ).getAsJsonArray();
+
+                        for ( JsonElement access : userAccesses )
+                        {
+                            if ( access.getAsJsonObject().get( "userUid" ).getAsString().equals( user.getUid() ) &&
+                                access.getAsJsonObject().get( "access" ).getAsString().matches( "..r.*" ) )
+                            {
+                                hasDataRead = true;
+                            }
+                        }
+
+                        if ( !hasDataRead )
+                        {
+                            for ( JsonElement access : userGroupAccess )
+                            {
+                                if ( user.getGroups()
+                                    .contains( access.getAsJsonObject().get( "userGroupUid" ).getAsString() ) &&
+                                    access.getAsJsonObject().get( "access" ).getAsString().matches( "..r.*" ) )
+                                {
+                                    hasDataRead = true;
+                                }
+                            }
+                        }
+                    }
+
+                    if ( hasDataRead )
+                    {
+                        dataRead.get( entry.getKey() ).add( obj.getAsJsonObject().get( "id" ).getAsString() );
+                    }
+                    else
+                    {
+                        noDataRead.get( entry.getKey() ).add( obj.getAsJsonObject().get( "id" ).getAsString() );
+                    }
+
+                } );
+            }
+        } );
+
+        user.setDataRead( dataRead );
+        user.setNoDataRead( noDataRead );
+    }
+
+    @ParameterizedTest
+    @ValueSource( strings = { "O2PajOxjJSa", "aDy67f9ijOe", "CKrrGm5Be8O", "Lpa5INiC3Qf", "GTqb3WOZMop" } )
+    public void testUserDataAndOrgUnitScopeReadAccess( String userUid )
+    {
+        User user = users.stream()
+            .filter( _user -> _user.getUid().equals( userUid ) )
+            .findFirst()
+            .orElseThrow( () -> new RuntimeException( "User UID not found for test" ) );
+
+        new LoginActions().loginAsUser( user.getUsername(), user.getPassword() );
+
+        QueryParamsBuilder queryParamsBuilder = new QueryParamsBuilder();
+        queryParamsBuilder.addAll( "ouMode=ACCESSIBLE", "fields=*" );
+        ApiResponse response = teiActions.get( "/", queryParamsBuilder );
+
+        response.validate().statusCode( 200 );
+
+        JsonObject json = response.getBody();
+
+        assertTrue( json.has( _TEIS ) );
```

Comment:
For better reporting, I would use rest-assured validation here. In that case, assertion failure here will be displayed with request, response and more meaningful message other than "Assertion failed. Expected true, but was false". 

It would be something like `response.validate().statusCode(200).body(_TEIS', is(notEmptyArray()))`